### DISCUSSION TITLE
How to access data partition created with repart?

### DISCUSSION THREAD

**Discussion by danielbisar** [systemd/mkosi#2353] [Q&A] [answered]
Host: Debian 12
mkosi v0.20.2 (git) + https://github.com/systemd/mkosi/pull/2347
Target: Ubuntu focal

Lets say i have a mkosi.repart/20-data.conf with the following content:

```
[Partition]
Type=linux-generic
Format=ext4
SizeMinBytes=1G
SizeMaxBytes=1G
```

On boot the partition is also shown in the overview before systemd starts.

This is just the start of a more complex setup. So there are a few questions:
- there is no fstab?
- context: system booted with "mkosi boot": how do i mount the additional partition (lsblk show loop5 -> loop5p3), if i try mount -o loop loop5p3 /mnt it says permission denied; and I guess i should probably create some systemd mount files?
- mount does not seem to have it mounted

- access to internet does not work in mkosi.postinst, event though [Packages] -> WithNetwork=yes is set.

**Comment by daandemeyer** [comment]
> there is no fstab?

I am adding fstab support in https://github.com/systemd/systemd/pull/30636. Until that lands, you have to write /etc/fstab yourself (which is complicated unless you use fixed UUIDS)

> context: system booted with "mkosi boot": how do i mount the additional partition (lsblk show loop5 -> loop5p3), if i try mount -o loop loop5p3 /mnt it says permission denied; and I guess i should probably create some systemd mount files?

You cannot mount additional partitions from within a systemd-nspawn containers. These have to be mounted on the host and passed into the container. Maybe systemd-nspawn should look at fstab, but this is something related to systemd, not mkosi, so I'd suggest opening an issue in the systemd repo.

> mount does not seem to have it mounted

There's no fstab, and systemd-nspawn doesn't automatically mount generic data partitions

> access to internet does not work in mkosi.postinst, event though [Packages] -> WithNetwork=yes is set.

I'll need a lot more information that that to be able to debug this

**Reply by danielbisar** [reply, marked as answer]
Ok, thatnk you. I guess I missunderstood the concept a bit.

What I came up with is the following:

30-data.conf
```
[Partition]
Type=linux-generic
Format=ext4
CopyFiles=/data:/
SizeMinBytes=1G
SzeMaxBytes=1G
```
    1. I create the data partition after the root, makes the order of loopback devices created more logical (1 = efi, 2 = root, 3 = data). I am wondering, since I thought that creating a the conf file like I had, that the content of /data will be deployed into the new partition and the content of the /data folder inside the root will be empty afterwards; the order root vs data doesn't seem to matter for that)
    2. create the image with mkosi
    3. on the host: create loop devices for each partition with `kpartx -av ./image.raw`
    4. mount the root partition and delete the content of /data
    5. mount the other partitions (in this case: /efi and /data)
    6. boot using systemd-nspawn -b --directory=./my_root
    7. umount via relative paths ./my_root/data, .../efi, ../
    8. kpartx -dv ./image.raw

I am writing you this for two reasons:
1. I am suspecting I am doing something wrong OR there might be a bug in mkosi or both....
2. Maybe you didn't know about this kpartx thing; isn't without sudo, but at least a way

I probably will implement some kind of script that will parse the output of kpartx and automount at least for simple cases like (root, efi, additional parts) for our CI.
